app/ide-desktop/lib/dashboard/src/authentication/src/authentication/cognito.ts

```diff
@@ -6,26 +6,143 @@
 import * as amplify from '@aws-amplify/auth'
 import * as cognito from 'amazon-cognito-identity-js'
 import * as results from 'ts-results'
+import * as app from '../components/app'
 
 import * as config from './config'
 
+// =================
+// === Constants ===
+// =================
+
+/** A list of errors thrown by the Amplify library that we catch and handle ourselves.
+ *
+ * Not all errors are caught and handled. Any errors not listed here are allowed to propagate up.
+ *
+ * Errors are grouped by the AWS Amplify function that throws the error (e.g., `signUp`).
+ *
+ * Each error must include an `internalCode` field, which is the code that the Amplify library uses
+ * to identify the error.
+ *
+ * Some errors also include an `internalMessage` field, which is the message that the Amplify
+ * library associates with the error. This field is used to distinguish between errors that have the
+ * same `internalCode`.
+ *
+ * Amplify reuses some codes for multiple kinds of errors. In the case of ambiguous errors, the
+ * `kind` field provides a unique string that can be used to brand the error in place of the
+ * `internalCode`, when rethrowing the error. */
+const KNOWN_ERRORS = {
+    /** Errors specific to the `currentSession` function. */
+    currentSession: {
```

Comment:
is there a benefit to namespacing errors like this? not that it's a bad thing, but:
- i haven't really seen things done this way before
- using these errors does become longer...

Reply:
maybe it would be useful to do something like this instead:
```ts
const CODE_TO_ERROR = {
  UsernameExistsException: {
    kind: 'UsernameExists'
  },
  ...
}
```

Reply:
doing it this way doesn't feel like it'd help for `intoXErrorOrThrow`, unfortunately... but it may still make the code a bit cleaner

Reply:
hm. another alternative is to split the big object into small objects for each endpoint. that way you can move them to the appropriate section

(i do believe there's nothing in the style guide saying all constants must be in a single section at the top)

Reply:
@indiv0 add short description of these reasons to the PR. I had the same questions as @somebody1234 and the reasoning is missing from docs :)

Reply:
> (i do believe there's nothing in the style guide saying all constants must be in a single section at the top)

@somebody1234 it might not be in the style guide but IIRC that's how the constants are usually handled.

Reply:
i believe it's largely because most other files only really have one section. in this case because afaict the only thing using them is the intoXError, it might be nice to have this constant + XErrorKind + XError all in one place

Reply:
Yeah alright that sounds good. I moved them together to centralize the error handling logic, but there's no point in that since the error handling logic isn't really generalizeable, since Amplify is so inconsistent.